### ISSUE TITLE
[normal] Implement isNFC()

### ISSUE THREAD

**Issue by behnam** [behnam/rust-unic#20]
Many APIs need to check input for being NFC, and it's something that can be done pretty fast for a majority of cases.

Ref: http://www.unicode.org/reports/tr15/#Detecting_Normalization_Forms

**Comment by CAD97** [comment]
isNFC, isNFKC, isNFD, and isNFKD rely on derived Quick_Check properties from [DeriedNormalizationProps]. If I'm reading tr44's [Decompositions_and_Normalization] information correctly, the intent is that these derived properties should be stored by the library and used by these algorithms in order to

> bypass the more time-consuming call to run the complete Unicode Normalization Algorithm on the input string.

Do we want to include these properties? Where should they go? `unic::ucd::normal::quick_check` seems the most logical position to me.

[DerivedNormalizationProps]: http://unicode.org/reports/tr44/#DerivedNormalizationProps.txt
[Decompositions_and_Normalization]: http://unicode.org/reports/tr44/#Decompositions_and_Normalization

**Comment by behnam** [comment]
Right. Agreed that having the property under `unic::ucd::normal` as a separate crate is the best option. Although it comes with the cost of making it public, which is not encouraged, but I think it's fine, because it's a low-level create.

To stay consistent with the UAX#44 recommendation to not put `Quick_Check` in a public API, we can just not expose it in the higher-level components (`unic::ucd::normal` and `unic::ucd` proper).

And then the string-level functionality would go under `unic::normal`, next to `toNFC()` implementation.

And regarding the name, I think `is_nfc()` would be clear.

About the existing methods, I think we need to rename `nfc()` (and others) to `to_nfc()`, to go along better with Unicode names AND follow the API guideline about conversion methods.

What do you think?

**Comment by CAD97** [comment]
I believe since the current `StrNormalForm::nfd` et. al. are conversion methods, the naming convention of `to_nfc` et. al. is desired.

And doesn't it say _to_ include `Quick_Check`?

> The Quick_Check property values are recommended for exposure in a public library API which supports Unicode character properties, [...]

Rather it's Decomposition_Mapping, Composition_Exclusion, and Full_Composition_Exclusion which I read as being intended private to the implementation rather than exposed. Since we're implementing the algorithms in `unic::normal` and the properties in `unic::ucd::normal`, They'd have to be pub in `unic-ucd-normal` (or whatever sub-crates we end up using or not) but not pub in `unic-ucd`, which is the api for accessing UCD properties.

**Comment by behnam** [comment]
Oops! You're right. `Quick_Check` is indeed recommended to be exposed. So, yeah, we can even expose its types in `unic-ucd`.

**Comment by CAD97** [comment]
@behnam This just popped into my head: the rename should be `unic_normal::StrNormalForm::nfd` => `unic_normal::StrNormalForm::into_nfd` since it consumes `self`. It could, however, not consume `self` on `str`...

**Comment by behnam** [comment]
Right. And I'm not sure why it's consuming the source. Could be intentional, just to prevent accidental usage after `nfc()` is available.

If there's no impl reason to keep it consuming, I think we better open it up to just borrow, and rename it to `to_nfc()`, which is what this functionality is normally called.

And, we will hve `is_nfc()` next to it, and iff `is_nfc()`, `to_nfc()` would be transparent and very cheap, otherwise expensive.

**Comment by CAD97** [comment]
I believe the reason `nfc()` is consuming right now is the `impl<I: Iterator<Item=char>> StrNormalForm<I> for I`. It makes sense to offer that, such that e.g. a stream of data from the internet (after being passed through a bytestream -> charstream combinator) can be converted into normal form. I think we'll want to offer two trait interfaces, one that takes a `&str` and returns a union of `str::Iter` or `De/Compositions`, and one that takes a `impl Iterator<Item=char>` and returns a `De/Compositions`.

**Comment by behnam** [comment]
Agreed that we probably need two (or more) sets of API for these functionalities. Working on `&str` can be very different, performance-wise and borrow-wise, from char-stream.

Right now we have NFC in `unic-normal`, `to_ascii`/`to_unicode` in `unic-idna`, some steps of `BidiInfo::new()` in `unic-bidi`, which all can share a similar API.

Related issue: https://github.com/behnam/rust-unic/issues/6

I'm not sure if right now is a good time to get to the API design of that level. I think should first add a couple of more components, specially segmentation, before we tackle that.

What do you think?

**Comment by CAD97** [comment]
Do we want to push this for UNIC 0.6 or delay it until later when it's the right time to be designing API design at this level? I could pretty easily do the legwork to expose the `Quick_Check` properties if we want to get this done soonish™️. (On a similar note, would it be worth it to bring in [`tribool`](https://crates.io/crates/tribool)? It seems like may Unicode algorithms use this [three-valued logic](https://en.wikipedia.org/wiki/Three-valued_logic).)

**Comment by behnam** [comment]
Would be great to stark exposing `Quick_Check`, now that we have `Char*` types in action. I think `isNFC()` can give a good boost to many applications, so worth prioritizing.

> On a similar note, would it be worth it to bring in tribool? It seems like may Unicode algorithms use this three-valued logic.

Rust compiler is expected to get a boost for `Option<bool>` and other similar types, dropping the need to create custom types for the same concept. Because of that, I would say it's not a priority, unless someone really needs it. If we switch to `tribool` or similar external types, then we will have to switch back at some point, which is extra work, again.
